### PR TITLE
inputstream: add API version check

### PR DESCRIPTION
**Pull request by FernetMenta**
adds API version check for inputstream addons

@mapfau /cc

### CODE REVIEW

**Comment by mapfau** [comment]
Looks good, thx

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
Android build error unrelated

File: xbmc/addons/InputStream.cpp [hunk 1]
```diff
@@ -69,6 +69,18 @@ CInputStream::CInputStream(const AddonProps& props,
...
+bool CInputStream::CheckAPIVersion()
+{
+  std::string dllVersion = m_pStruct->GetApiVersion();
+  if (dllVersion.compare(INPUTSTREAM_API_VERSION) != 0)
+  {
+    CLog::Log(LOGERROR, "CInputStream::CheckAPIVersion - API version does not match");
+    return true;
+  }
+
+  return true;
```
Comment (da-anda) [inline xbmc/addons/InputStream.cpp]: why are you in both cases returning `true`? Doesn't make much sense.

Comment (FernetMenta) [inline xbmc/addons/InputStream.cpp]: oops, this is obviously not correct

**Comment by MilhouseVH** [comment]
When a older version of `inputstream.mpd` is installed (eg. v1.2.0, though probably not specific to this particular inputstream add-on) and the user upgrades to a version of Kodi with that includes this API check, the newer version of Kodi will crash on startup (see post [here](https://github.com/LibreELEC/LibreELEC.tv/pull/641#issuecomment-241618163)).

This currently makes it impossible for inputstream users to upgrade Kodi without manual intervention (ie. deleting inputstream add-ons, and then re-installing updated versions of the same add-ons).

Is there a simple fix to prevent a new version of Kodi from crashing when an older add-on is installed, as this is going to be a fairly common scenario.

**Comment by FernetMenta** [comment]
ideally cpluff would care about versions but this is not implemented yet for binary addons. unfortunately I can't think of any simple fix for this. taking into account that inputstream api comes with v17 only alpha users are affected.

**Comment by MilhouseVH** [comment]
It's a little surprising that a previously released add-on will be able to crash Kodi in this way.

Although it's currently affecting only alpha users (and we have quite a few!) it's likely to be an issue even with users upgrading from an alpha build to a full Kodi release build unless there's a fix of some kind added to Kodi. Upgrading the current add-on is a bit of a chicken/egg situation, as I'm not sure if the new add-on (ie. with API check support) will work with the older non-API Kodi builds.

Is it not possible to simply catch the error (and possibly disable the add-on), rather than segfault?

**Comment by FernetMenta** [comment]
Nope, this is an incompatibility of ABI. The only way to prevent from segfault would be to have cpluff check version is advance and not load the shared library. But this is not done by our addon system. C has no capability of catching errors.

Alpha users have to be aware that those type of things can happen with developer versions.

**Comment by garbear** [comment]
@MilhouseVH I've run into a similar problem, Kodi crashing on startup due to an add-on being present. I tracked it down to a circular dependency on initialization. Can you give 230c1d8 a try?

**Comment by MilhouseVH** [comment]
@garbear thanks, I'll give it a try (might take a few hours before I can confirm)

**Comment by MilhouseVH** [comment]
@garbear unfortunately your patch hasn't made any difference, Kodi continues to crash on startup. http://sprunge.us/Vdhj

**Comment by mapfau** [comment]
@FernetMenta shouldn't we check for the kodi.inputstream version always on load time?

It's 1.0.4 with incompatible versions and was bumped to 1.0.5. There is no check at all against this version when loading inputstream.xy addons.

**Pull request merged**